transformer_engine/pytorch/module/base.py

```diff
@@ -1179,6 +1182,48 @@ def _validate_name(self):
             )
             self.name = f"Layer_{TEDebugState.get_layer_count()}"
 
+    def _check_weight_tensor_recipe_correspondence(self) -> None:
+        """
+        Verify that the weight tensor types match their corresponding recipe type.
+        This is invoked in the forward().
+
+        This establishes a 1:1 correspondence between recipe types and tensor types:
+        - DelayedScaling → Float8Tensor
+        - Float8CurrentScaling → Float8Tensor
+        - MXFP8BlockScaling → MXFP8Tensor
+        - Float8BlockScaling → Float8BlockTensor
+
+        Example case to check: recipe is DelayedScaling (DelayedScaling is set in fp8_autocast()),
+        but the weight tensor is MXFP8Tensor (MXFP8BlockScaling is set in fp8_model_init()).
+        """
+        if not self.fp8 and not self.fp8_calibration:
+            return
+        if not hasattr(self, "weight_names") or not self.weight_names:
+            return
+
+        recipe = self.fp8_meta["recipe"]
+        expected_tensor_class = None
+        if recipe.delayed() or recipe.float8_current_scaling():
+            expected_tensor_class = Float8Tensor
+        elif recipe.mxfp8():
+            expected_tensor_class = MXFP8Tensor
+        elif recipe.float8_block_scaling():
+            expected_tensor_class = Float8BlockwiseQTensor
+        else:
+            raise RuntimeError(f"Unsupported recipe type: {recipe.__class__.__name__}")
+
+        weight_tensors = [getattr(self, name) for name in self.weight_names]
```

Comment:
@timmoon10 I vaguealy recall that getattr from nn.Module was slow and that you created a faster function for it at some point, do you remember the details?